Symfony events added for login failed and public link accessed

sharidas opened the pull request:
New symfony events added when:
a) login attempt is failed
b) public link share is acced

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
This change comprises of emitting symfony events when :
- A user fails to login
- A public link share is accessed.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- https://github.com/owncloud/core/issues/31462
- https://github.com/owncloud/core/issues/31465

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
The login failure can be logged using the event triggered. Which would be helpful for the administrators to know which user tried to login and failed. Public link share event would also help in the same way.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- For the public link share access, tested in the debug mode. When the link is accessed, the control reaches the dispatcher.
- For the login failed tested using debug mode.
- Updated the tests for the same.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
Symfony events added for login failed and public link accessed

New symfony events added when:
a) login attempt is failed
b) public link share is acced

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented on apps/files_sharing/lib/Controllers/ShareController.php:
```diff
@@ -223,6 +229,16 @@ protected function emitAccessShareHook($share, $errorCode = 200, $errorMessage =
 		]);
+
+		if ($share instanceof \OCP\Share\IShare) {
+			$cloneShare = clone $share;
```
If clone is not to be used, let me know I will remove it.

karakayasemi commented:
As far as I can see we have four different login types in UserSession, and UserSession is already emitting a failedLogin events for two login types. 
https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L494
https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L624

Probably, this PR leads to emit two different failedLogin events for loginWithPassword and loginWithApache. IMHO, we should unify these events. 

In my pov, we can remove event signal emitting logic from UserSession and use this PR, or add failedLogin event for the other missing login types in UserSession.

PVince81 commented on apps/files_sharing/lib/Controllers/ShareController.php:
```diff
@@ -223,6 +229,16 @@ protected function emitAccessShareHook($share, $errorCode = 200, $errorMessage =
+				['shareObject' => $cloneShare, 'errorCode' => $errorCode,
+					'errorMessage' => $errorMessage]);
+
+			$this->eventDispatcher->dispatch('share.linkaccess', $publiShareLinkAccessEvent);
```
missing "c"

PVince81 commented:
@sharidas I'd have preferred separate PRs or at least separate commits since the two events are not related at all (or are they?). It's ok to leave it as one now.

butonic commented on apps/files_sharing/tests/Controllers/ShareControllerTest.php:
```diff
@@ -258,6 +263,13 @@ public function testAuthenticateInvalidPassword() {
 
+		$calledShareLinkAccess = [];
+		$this->eventDispatcher->addListener('share.linkaccess',
+			function (GenericEvent $event) use (&$calledShareLinkAccess) {
```
ou are not even using `$calledShareLinkAccess` ... get rid of the use.
also, stop using & everywhere. there are only rare occasions when you need it: http://php.net/manual/en/language.references.whatdo.php

sharidas replied:
`$calledShareLinkAccess` is used from https://github.com/owncloud/core/pull/31467/files#diff-d1ba03f4d58be19f7cb3af8c272e150eR288 to https://github.com/owncloud/core/pull/31467/files#diff-d1ba03f4d58be19f7cb3af8c272e150eR296

If I remove the use of `&` then by the time when test tries to use assertions the `$calledShareLinkAccess` would be empty, just like https://github.com/owncloud/core/pull/31467/files#diff-d1ba03f4d58be19f7cb3af8c272e150eR266

sharidas commented:
@PVince81 
> did you address https://github.com/owncloud/core/pull/31467#issuecomment-389859926 ?

The issue raised in this comment was because my earlier version of the PR used LoginController.php to dispatch the events. But in this version I have removed the change from LoginController.php to `User/Session.php`. Hence I have addressed the comment by having the dispatchers related to login failures in `User/Session.php`.

PVince81 commented on apps/files_sharing/lib/Controllers/ShareController.php:
```diff
@@ -223,6 +229,16 @@ protected function emitAccessShareHook($share, $errorCode = 200, $errorMessage =
+
+		if ($share instanceof \OCP\Share\IShare) {
+			$cloneShare = clone $share;
+			$publiShareLinkAccessEvent = new GenericEvent(null,
```
s/publi/public/

PVince81 reviewed (approved):
Looks fine :+1:

sharidas commented:
Backport PR: https://github.com/owncloud/core/pull/31584
